File: src/main/java/io/github/takenoha/towerdefense/paper/PlayerRecoveryGuard.java
package io.github.takenoha.towerdefense.paper;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

/** Main-thread lifecycle gate used while a player's voucher state is being reconciled. */
final class PlayerRecoveryGuard {
    private final Set<UUID> playerIds = new HashSet<>();

    void begin(UUID playerId) {
        playerIds.add(playerId);
    }

    void complete(UUID playerId) {
        playerIds.remove(playerId);
    }

    boolean isGuarded(UUID playerId) {
        return playerIds.contains(playerId);
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/PlayerRecoveryGuard.kt
package io.github.takenoha.towerdefense.paper

import java.util.HashSet
import java.util.UUID

/** Main-thread lifecycle gate used while a player's voucher state is being reconciled. */
class PlayerRecoveryGuard {
    private val playerIds: MutableSet<UUID> = HashSet()

    fun begin(playerId: UUID) {
        playerIds.add(playerId)
    }

    fun complete(playerId: UUID) {
        playerIds.remove(playerId)
    }

    fun isGuarded(playerId: UUID): Boolean = playerIds.contains(playerId)
}
